### DISCUSSION TITLE
Import "cv2.ximgproc" could not be resolved

### DISCUSSION THREAD

**Discussion by mpizenberg** [microsoft/pyright#5236] [Q&A] [answered]
Hi, I’m currently using pyright for a project using OpenCV (with contrib extension), but pyright does not seem to be able to pick up the existence of that `cv2.ximgproc` module.

To reproduce the issue, you can use the following steps:

1. create a new conda env, activate it
2. run `pip install opencv-contrib-python`
3. create a new `example.py` file with the line `import cv2.ximgproc`
4. run `pyright .`

```sh
Found 1 source file
pyright 1.1.300
C:\Users\MatthieuPIZENBERG\temp\python\opencv\example.py
  C:\Users\MatthieuPIZENBERG\temp\python\opencv\example.py:2:8 - error: Import "cv2.ximgproc" could not be resolved (reportMissingImports)
1 error, 0 warnings, 0 informations
```

![image](https://github.com/microsoft/pyright/assets/2905865/d1a82638-c0ce-4f05-83b6-8b1d47827e0a)

Yet running `python example.py` works fine and the import is successful. This seems like a bug but I preferred asking here first. Should I report this in an issue? Any idea how to solve that?

My pyright version is 1.1.300
I’m on Windows

**Comment by erictraut** [comment, marked as answer]
The `cv2` library is an untyped library — i.e. it ships with no type annotations or type stubs. Furthermore, it's written primarily in a language other than Python, so pyright's ability to extract type information out of it is very limited.

Looking at the top-level Python module in cv2, I don't see any symbols named `ximgproc`. I also don't see a submodule named `ximgproc` in the library. If this works at runtime, it probably means this symbol is being dynamically injected into the namespace of the top-level module. That's something that a static type checker wouldn't have knowledge of.

If `cv2` is important to you, I recommend contacting the maintainers of the library and asking them to provide type information (packaged type stubs). This is a popular library, and we get lost of questions about it. The more people who ask for this functionality, the more likely it will get done.

Pyright is working as intended here, so there's no bug to fix.

Your short-term options include:
1. Disable the `useLibraryCodeForTypes` option. This will cause pyright to treat all symbols imported from `cv2` as type `Unknown`. It will make the error disappear, but you will also lose all completion suggestions and other type checking for the library.
2. Use a `# type: ignore` or `# pyright: ignore` comment to suppress the error on that line.
3. Write your own (minimal) type stubs for `cv2` and place them in the `typings/cv2/` subdirectory within your project. You could start with a stub that simply defines a module-level `__getattr__` function with an `Any` return type and then fill in specific functions, classes, and submodules as you see fit. This is a lot of work, but it will give you the best results.

**Comment by mpizenberg** [comment]
Thanks for the detailed answer and suggestions!
